experimental/list_formatter/src/list_formatter.rs

```diff
@@ -46,86 +48,138 @@ impl ListFormatter {
         Ok(Self { data, width })
     }
 
-    fn format_internal<'c, B>(
+    fn format_internal<'c, W: Writeable, S: ?Sized>(
         &'c self,
-        values: &[&str],
-        mut builder: B,
-        append_value: fn(B, &str) -> B,
-        append_literal: fn(B, &str) -> B,
-    ) -> B {
+        values: &[W],
+        sink: &mut S,
+        write_w_value: fn(&mut S, &W) -> fmt::Result,
```

Comment:
Nit: it's more canonical in Rust to use `Fn`, `FnOnce`, or `FnMut`.

Reply:
This is a function pointer though. I don't want to introduce generic type bounds to make this work with closures if it's not required.

Reply:
I don't have a strong preference here, but if you used `impl Fn`, then you wouldn't have to add any more generics at the top-level. 

```diff
-        write_w_value: fn(&mut S, &W) -> fmt::Result,
-        write_m_value: fn(&mut S, &str) -> fmt::Result,
-        write_literal: fn(&mut S, &str) -> fmt::Result,
+        write_w_value: impl Fn(&mut S, &W) -> fmt::Result,
+        write_m_value: impl Fn(&mut S, &str) -> fmt::Result,
+        write_literal: impl Fn(&mut S, &str) -> fmt::Result,
```

Not sure if this would have a negative impact on code size.